### ISSUE TITLE
Restrict access to resource metadata

### ISSUE THREAD

**Issue by genkimaps** [genkimaps/gis-hub#86]
A custom version of the ckan restricted extension is used on the GIS Hub to prevent unauthorized users from getting access to a resource's metadata page.  However, any user with the direct URL to that resource can view the metadata, including the download link and map preview. 

This issue is to disable display of any resource metadata for unauthorized users, even if they know the URL. 

Will try to adapt code from https://github.com/genkimaps/ckanext-restricted/blob/master/ckanext/restricted/templates/package/snippets/resource_item.html, which is where the "Request Access" button is displayed instead of a link to the resource. 

I think this can be done using:

`{% set authorized = h.check_access('resource_show', {'id': res.id, 'resource': res }) %}`

and display the metadata only if authorized. 

Related to: https://github.com/genkimaps/mapserver/issues/3

**Comment by genkimaps** [comment]
@cole-fields  Did we already create an issue for this? I don't think we ever did, we just discussed it by email.

**Comment by cole-fields** [comment]
@genkimaps no we never created an issue for this, thanks for catching. working on some code that will use the resource patch function in the ckanapi python code. the updated code is to go through list of resources and updated allowed users for each one. this will prevent unauthorized users from browsing to the layer but not prevent access if they already have the url. pushing to a separate branch here: https://github.com/genkimaps/hub-geo-api/commit/2048752f252ad88cea0ca5c5bdce30ed32cb0281

**Comment by genkimaps** [comment]
@cole-fields  I updated the resource_read template in this branch:
https://github.com/genkimaps/gis-hub/tree/restrict-resdata

Even if a user has the URL directly to a resource, if not authorized, they won't see any of the custom metadata.  Views and Go To Resource button also hidden.  Please try this with your dummy account and let me know if it works for you.

**Comment by cole-fields** [comment]
@genkimaps this is great - i think this is exactly what we needed, although i don't really understand the code/how that works. seems to be functional though, judging from my dummy account:

![image](https://user-images.githubusercontent.com/52669482/93138244-8d54ed00-f693-11ea-8ba9-458af2bf35f2.png)

**Comment by cole-fields** [comment]
@genkimaps the only workaround that is still a concern is this:
 - user does not have access to resource
 - user can run package_show https://www.gis-hub.ca/api/3/action/package_show?id=env-layers-wcvi-20-m to get the metadata including the web map preview link
 - with that URL, the user has access to the preview

I was able to do this with my dummy account. This is a fairly unlikely scenario; however, maybe there is a simple solution.

**Comment by cole-fields** [comment]
@genkimaps since this is working, we should merge it with master. i was going to myself (the branch is missing some shell scripts for database backup that are in master), but there was a conflict that i did not want to touch since gis hub server is currently running on restrict-resdata branch.

**Comment by genkimaps** [comment]
@cole-fields  In theory, it should be possible to globally restrict access to the CKAN API (https://www.gis-hub.ca/api) at the webserver level, i.e. nginx.  However, I am not sure whether this will interfere with CKAN's internal authorization, which uses a per-user API key.  My guess is it won't interfere, but I won't know until I try.

**Comment by cole-fields** [comment]
@genkimaps that sounds like the best solution, except that I use the API in a few scripts that i run from my local machine rather than on the server. So in that sense, it may not be ideal. Unless there's a way to allow exceptions.

I guess that leads to a decision point of: 
a) increasing the security level by globally restricting access to ckan api
b) leaving as is with the slight security concern, but api is still accessible

I would lean towards option a - but even thinking about the tool you provided to Nick for emergency response planning where the user drags the text file onto the executable file and it downloads data from the hub - that functionality would no longer be supported, correct?

**Comment by genkimaps** [comment]
@cole-fields Yes, option A would break existing tools such as the data download tool for Nick.  Would have to recompile that tool.  If you use the same version of the API that I have in the hub-geo-api repo, you should only need to update one file, so it might not have as much impact as you expect.

**Comment by cole-fields** [comment]
@genkimaps ok, i guess if the tool will work after being recompiled, then that sounds good. not quite sure i understand. does restricting access to the API only mean through the web interface (https://www.gis-hub.ca/api)? but I'd still be able to access in via the command line and through Python code run on my desktop, outside of the gis-hub server? Or would i only be able to do that on the server?

**Comment by genkimaps** [comment]
@cole-fields This will affect anyone using the API, whether it's through Python code, a web browser, or anything else.  On a web browser, you'll be prompted for a password.  If using the Python requests module, you need to include a password with the command, which should be loaded from environment vars.  This is why we need to make sure that there's only one copy of the ckanapi.py floating around.

**Comment by genkimaps** [comment]
@cole-fields 
I tried this now, and the problem with restricting access to the entire API is that just a basic page load of any page in CKAN requires hidden calls to the API (made by jQuery).  It will also block things like autocomplete.  Can see these from the Network tab of the developer console:

![image](https://user-images.githubusercontent.com/15989988/94358646-7d023080-00dd-11eb-9a65-9697e9790fa6.png)

However, this might work if I restrict access only to the package/resource sections of the API.

**Comment by genkimaps** [comment]
@cole-fields  I limited the restriction to cover only the package/resource endpoints in the API.  This worked, and I tested it with my copy of ckanapi.py.  I disabled it for now, because I want to make sure that we are all using the correct version of ckanapi.py before moving ahead with this.

**Comment by cole-fields** [comment]
@genkimaps thanks for the explanation. I'll have to review this more this week, but that sounds like a good solution. I've merged the branch from the pull request into the master branch of hub-geo-api and so i should be working with the most current version of ckanapi.py

**Comment by genkimaps** [comment]
@cole-fields 
The change to ckanapi.py has been made here:
https://github.com/genkimaps/hub-geo-api/commit/22067494ee9e8e9b1f176d023fa02830f145da8d
There is now a required environment variable:  CKANAPI_HTTP_AUTH
You will need to get the latest version of master branch in the 'hub-geo-api' repo. 
I will separately send you the credentials for CKANAPI_HTTP_AUTH.  Please put this in your environment vars on your local PC (Please don't make any changes to environment vars on the GIS hub and mapserver, at least not yet).

**Comment by cole-fields** [comment]
@genkimaps thanks - i've pulled from the master branch and added the CKANAPI_HTTP_AUTH value in my environmental vars on my desktop. looking forward to seeing this in action as this seemed to be the last security-related issue we had

**Comment by genkimaps** [comment]
@cole-fields  Before I can make this change on the prod server, I need to update the gokit tool.  Also, we need to distribute the updated version of that tool to existing users.  I am aware of Nick Benoy and Matthias, but there may be others too.  Can you try to track down the other active users?

**Comment by cole-fields** [comment]
@genkimaps  I've sent it to a couple other folks outside of DFO - but i don't think it's important to send them a new copy unless they request it down the road. I imagine they got the data they were after and wouldn't need to use the tool again. Besides those users listed, myself, Kayleigh, and possibly Joanne. Typically other people are accessing via the Hub, but the tool will still be useful for us.

**Comment by genkimaps** [comment]
@cole-fields  Restricted access to the `resource` and `package` API paths in CKAN is now enabled.  Please see the following list of changes that were made: 

1. Enabled the location block in the nginx config file, which requires a user/pass for `resource` and `package`.  https://github.com/genkimaps/gis-hub/commit/a9b10274de786ba71b69ac879c315f6da019f4b8.  Restarted nginx service only (no need to restart CKAN). 
2. Added the `CKANAPI_HTTP_AUTH` environment var to file `/etc/environment` on the gis-hub.ca server.  
3. Updated ckanpi.py in the hub-geo-api backend script, and restarted the service.  I also merged changes from the restricted_meta branch: https://github.com/genkimaps/hub-geo-api/commit/e61fe0e1721c13f9891cb90128c5a90b75b1da41
3. Added the `CKANAPI_HTTP_AUTH` environment var to file `/home/dfo/mapserver.env` on maps.gis-hub.ca
4. Updated ckanpi.py in the mapserver code, taking only the relevant portions (some functions, such as renaming users, are never used on mapserver).  https://github.com/genkimaps/mapserver/commit/e7b266e15b17c8b242cef5295041d4d96643dcbd.  Restarted the Python-RQ listener on mapserver. 

Still having some issues with this.

**Comment by genkimaps** [comment]
@cole-fields  Unfortunately I had to revert this change, and disable the additional password protection for the CKAN API.  The reason is that the API authorization field required for CKAN uses the same HTTP header as the HTTP Basic authentication header.  Both of them set a header called 'Authorization' 

I might possibly be able to get this to work using header rewrites in nginx, but I don't feel up to trying that right now.  
https://stackoverflow.com/questions/12431496/nginx-read-custom-header-from-upstream-server/30879181#30879181

Basically in this line:
`r = requests.post(url, headers=ghub_headers, json=data, auth=ckan_http_auth)` 
The statement `auth=ckan_http_auth` overwrites the authorization headers set in `headers=ghub_headers`

**Comment by genkimaps** [comment]
@cole-fields  This change is now done, I set it up so that the only change needed was in the nginx conf file.  It requires the same API key as in the ckanapi.py file.  

https://github.com/genkimaps/gis-hub/commit/b0d69c03c512efd475e1d208af76370d66419cee

This is a copy of the nginx conf file.  The real one contains the API key itself, so it should not be in version control.  Nginx doesn't support environment vars, so I had to hardcode it.  The nginx file that is actually being used is:  `/home/tk/ckanext-dfo/conf/dfo.gishub.nginx.conf`

Because of this, you no longer need CKANAPI_HTTP_AUTH in your environment vars.  You don't have to remove it, but you might want to do so to keep things clean.

**Comment by genkimaps** [comment]
@cole-fields  A note on this change: if you try to access any of the restricted API pages from a browser, you'll get an access denied message.  This is by design.  If you want to access these URLs, you can use a browser extension such as the one below to add an HTTP header Authorization: XXX with the API key. 

https://chrome.google.com/webstore/detail/modheader/idgpnmonknjnojddfkpgkljpfnnfcklj

**Comment by cole-fields** [comment]
@genkimaps excellent! I think even though the potential user would have required some knowledge of the CKAN API to exploit the security issue, it's great that we're now just preventing that entirely. Nice work - I imagine that was a bit of trial and error to have it work as desired.

**Comment by cole-fields** [comment]
Unfortunately IT has disabled adding extensions to browsers, but I'll add this to my personal computer.

**Comment by genkimaps** [comment]
@cole-fields  If you do use the ModHeader extension, be sure to add a filter so that it only applies to the GIS-hub URL.  Otherwise you will inadvertently be sharing that key with every website!  

![image](https://user-images.githubusercontent.com/15989988/96082679-a224da80-0ef6-11eb-87c4-f5383aafaea8.png)

**Comment by genkimaps** [comment]
@cole-fields  This seems to be working now, closing the issue.

**Comment by cole-fields** [comment]
@genkimaps agreed - working well. the mod header extensions is also very handy, even if i can only use if on my personal laptop

**Comment by genkimaps** [comment]
@cole-fields Reopening this bug because the way this was done had some unintended consequences.  We have been checking for a specific CKAN API key in the request headers in nginx.  This was an ugly hack, but it worked.  The problem is that it restricts access to package_* and resource_* in the API, and we need this for the gokit script:  https://github.com/genkimaps/gokit.  

The previous version of gokit had the master CKAN API key (from an admin account) embedded in the script, which was a security issue.  I changed gokit to require the user to enter the API key from their account.  However, since the package_* endpoint is blocked, there is no way for the script to get the dataset download link.  

I think I have now fixed this in ckanext-restricted: https://github.com/genkimaps/ckanext-restricted/blob/b3b4567be600c73f9c73fab7cb4d708da8b8b651/ckanext/restricted/action.py#L220

ckanext-restricted was never designed to hide resource fields in the metadata, only to replace the link on the dataset page with a 'Request Access' button.  This was an oversight, but consistent with the CKAN philosophy of open data by default.   Now the API is also locked down, an unauthorized user sees something like this: 

![image](https://user-images.githubusercontent.com/15989988/101777085-4ca32d80-3b35-11eb-824e-8deb5c929a04.png)

This is a much better solution in my opinion, change is live in ckanext-restricted, branch hideresources.  I have removed the API header check in nginx.
